### PR TITLE
Swap chains, part 1

### PR DESCRIPTION
**Pull request by kainino0x**
Adds the swap chain interfaces to the API without changing the behavior of anything else. This includes the C APIs for applications to provide swap chain implementations. Also adds stub implementations on every backend.

### CODE REVIEW

File: src/backend/SwapChain.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    void SwapChainBase::Configure(nxt::TextureFormat format, uint32_t width, uint32_t height) {
+        this->format = format;
+        this->width = width;
+        this->height = height;
+    }
+
+    TextureBase* SwapChainBase::GetNextTexture() {
+        auto* builder = device->CreateTextureBuilder();
```
Comment (Kangz) [inline src/backend/SwapChain.cpp]: We should return an error if Configure wasn't called before.

File: src/backend/SwapChain.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        lastNextTexture = texture;
+        return texture;
+    }
+
+    void SwapChainBase::Present(TextureBase* texture) {
+        if (texture != lastNextTexture) {
+            device->HandleError("Tried to present something other than the last NextTexture");
+            return;
+        }
+
```
Comment (Kangz) [inline src/backend/SwapChain.cpp]: We should check the texture is in the present usage.

File: src/backend/metal/ShaderModuleMTL.mm [hunk 3]
```diff
@@ -29,6 +29,7 @@
         ExtractSpirvInfo(*compiler);
 
         std::string msl = compiler->compile();
+        printf("%s\n", msl.c_str());
```
Comment (Kangz) [inline src/backend/metal/ShaderModuleMTL.mm]: nit: stray printf

File: src/backend/metal/TextureMTL.mm [hunk 4]
```diff
@@ -73,7 +73,12 @@ MTLTextureType MetalTextureType(nxt::TextureDimension dimension) {
         mtlTexture = [mtlDevice newTextureWithDescriptor:desc];
     }
 
+    Texture::Texture(TextureBuilder* builder, id<MTLTexture> mtlTexture)
+        : TextureBase(builder), mtlTexture(mtlTexture) {
+    }
+
     Texture::~Texture() {
+        // TODO(kainino@chromium.org): Maybe don't release when using the native texture constructor?
```
Comment (Kangz) [inline src/backend/metal/TextureMTL.mm]: Or we could add a ref in the WSI constructor, maybe.

File: src/backend/opengl/TextureGL.cpp [hunk 5]
```diff
@@ -48,8 +48,18 @@ namespace opengl {
 
     // Texture
 
+    static GLuint GenTexture() {
```
Comment (Kangz) [inline src/backend/opengl/TextureGL.cpp]: nit: this can be in the anonymous namespace above.

File: src/backend/metal/TextureMTL.mm (hunk 4 above)
Comment (kainino0x) [inline src/backend/metal/TextureMTL.mm]: Oh, good idea. Works for this case, but what about the D3D12/GL backends?

**Pull request merged**